File: aegis-cpp/example/market_data.cpp
#include "aegis/client.hpp"

#include <atomic>
#include <chrono>
#include <csignal>
#include <iostream>
#include <memory>
#include <stop_token>
#include <thread>

// ---------------------------------------------------------------------------
// Handler
// ---------------------------------------------------------------------------

class MarketDataHandler : public aegis::ComponentHandler {
public:
    void on_configure(
        const std::string&              socket_path,
        const std::vector<std::string>& topics
    ) override {
        std::cout << "[MarketData] Configuring — socket=" << socket_path << "\n";
        // Connect to data stream here
    }

    void on_running(std::stop_token stop) override {
        std::cout << "[MarketData] Running — starting data worker\n";
        int tick = 0;
        while (!stop.stop_requested()) {
            ++tick;
            std::cout << "[MarketData] Tick #" << tick << " — processing data...\n";
            std::this_thread::sleep_for(std::chrono::seconds(5));
        }
        std::cout << "[MarketData] Data worker stopped\n";
    }

    void on_ping() override {
        std::cout << "[MarketData] PING received\n";
    }

    void on_shutdown() override {
        std::cout << "[MarketData] Releasing resources\n";
    }

    void on_error(const std::string& code, const std::string& message) override {
        std::cerr << "[MarketData] Error — code=" << code << " message=" << message << "\n";
    }
};

// ---------------------------------------------------------------------------
// Entry point
// ---------------------------------------------------------------------------

static aegis::Client* g_client = nullptr;

int main() {
    aegis::Config cfg;
    cfg.socket_path              = "/tmp/aegis_components.sock";
    cfg.session_token            = "YOUR_SESSION_TOKEN_HERE";
    cfg.component_name           = "market_data";
    cfg.version                  = "0.1.0";
    cfg.supported_symbols        = {"BTC/USDT", "ETH/USDT"};
    cfg.supported_timeframes     = {"1m", "5m"};
    cfg.requires_streams         = {"candles"};
    cfg.max_reconnect_attempts   = 10;

    auto client = std::make_unique<aegis::Client>(
        cfg,
        std::make_unique<MarketDataHandler>()
    );

    g_client = client.get();

    // Handle Ctrl-C
    std::signal(SIGINT,  [](int) { if (g_client) g_client->stop(); });
    std::signal(SIGTERM, [](int) { if (g_client) g_client->stop(); });

    try {
        client->run();
    } catch (const std::exception& e) {
        std::cerr << "Component stopped: " << e.what() << "\n";
        return 1;
    }
    return 0;
}

File: aegis-rust/examples/market_data.rs
//! Basic example: a component that connects to Aegis, receives market data,
//! and resets cleanly on session restart (REBORN).

use libaegis::{AegisError, Component, ComponentHandler, Config, Result};
use tokio_util::sync::CancellationToken;
use tracing::info;

// ---------------------------------------------------------------------------
// Handler
// ---------------------------------------------------------------------------

struct MarketDataHandler;

impl ComponentHandler for MarketDataHandler {
    async fn on_configure(&self, socket_path: String, topics: Vec<String>) -> Result<()> {
        info!("configure — socket={} topics={:?}", socket_path, topics);
        // Open the data stream socket here, e.g.:
        // let mut stream = libaegis::DataStream::connect(&socket_path, &component_id, &session_id).await?;
        Ok(())
    }

    async fn on_running(&self, shutdown: CancellationToken) -> Result<()> {
        info!("running — starting data worker");

        // Spawn the processing loop. It keeps running across session restarts —
        // REBORN does not cancel this token or call on_running again.
        // Reset per-run state in on_reborn() instead.
        tokio::spawn(async move {
            let mut interval = tokio::time::interval(std::time::Duration::from_secs(5));
            let mut tick = 0u64;
            loop {
                tokio::select! {
                    _ = shutdown.cancelled() => {
                        info!("data worker stopped");
                        break;
                    }
                    _ = interval.tick() => {
                        tick += 1;
                        info!("tick #{} — processing data...", tick);
                    }
                }
            }
        });

        Ok(())
    }

    async fn on_reborn(&self) {
        // Called when Aegis restarts the session (REBORN command).
        // Reset ALL state from the previous run here — counters, positions,
        // buffers, etc. on_running is NOT called again; the same task keeps
        // running and will receive data from the new orchestrator run.
        info!("reborn — resetting per-run state");
    }

    async fn on_ping(&self) {
        info!("ping");
    }

    async fn on_shutdown(&self) {
        info!("shutdown — releasing resources");
    }

    async fn on_error(&self, code: String, message: String) {
        tracing::error!("aegis error — code={} message={}", code, message);
    }
}

// ---------------------------------------------------------------------------
// Entry point
// ---------------------------------------------------------------------------

#[tokio::main]
async fn main() -> std::result::Result<(), Box<dyn std::error::Error>> {
    tracing_subscriber::fmt()
        .with_env_filter("libaegis=debug,market_data=info")
        .init();

    let socket_path   = std::env::var("AEGIS_SOCKET")
        .unwrap_or_else(|_| "/tmp/aegis-components.sock".into());
    let session_token = std::env::var("AEGIS_SESSION_TOKEN")
        .unwrap_or_default();

    let mut cfg = Config::new(socket_path, session_token, "market_data");
    cfg.version                = "0.1.0".into();
    cfg.supported_symbols      = vec!["BTCUSDT".into(), "ETHUSDT".into()];
    cfg.supported_timeframes   = vec!["1m".into(), "5m".into()];
    cfg.requires_streams       = vec!["aggTrades".into(), "klines".into()];
    cfg.max_reconnect_attempts = 10;

    let component = Component::new(cfg, MarketDataHandler);

    tokio::select! {
        res = component.run() => {
            if let Err(e) = res {
                eprintln!("component stopped with error: {e}");
                std::process::exit(1);
            }
        }
        _ = tokio::signal::ctrl_c() => {
            info!("ctrl-c — shutting down");
        }
    }

    Ok(())
}